cache.go

```diff
@@ -17,79 +17,90 @@
 //     2) The admission and eviction policy: this is the Policy interface.
-//     3) The bp-wrapper buffer: this is the ring.Buffer struct.
+//     3) The bp-wrapper buffer: this is the Buffer struct.
 //
 // All three of these components work together to try and keep the most valuable
 // key-value pairs in the hash map. Value is determined by the Policy, and
 // BP-Wrapper keeps the Policy fast (by batching metadata updates).
 type Cache struct {
-	data   store.Map
+	data   store
 	policy Policy
-	buffer *ring.Buffer
+	buffer *ringBuffer
+	notify func(string)
+	used   uint64
+	size   uint64
 }
 
 type Config struct {
-	CacheSize  uint64
-	BufferSize uint64
-	Policy     func(uint64, store.Map) Policy
-	Log        bool
+	// NumCounters is the number of counters to preallocate space for. This
+	// should correspond to the expected number of items in the Cache when it
+	// is full (when MaxCost is reached).
+	NumCounters uint64
+	// MaxCost is the cost capacity of the Cache.
+	MaxCost uint64
+	// BufferItems is max number of items in access batches (BP-Wrapper).
+	BufferItems uint64
+	// OnEvict is ran for each key evicted.
+	OnEvict func(string)
+	// Log is whether or not to Log hit ratio statistics (with some overhead).
+	Log bool
 }
 
 func NewCache(config *Config) *Cache {
+	if config.MaxCost == 0 && config.NumCounters != 0 {
+		config.MaxCost = config.NumCounters
+	}
 	// data is the hash map for the entire cache, it's initialized outside of
 	// the cache struct declaration because it may need to be passed to the
 	// policy in some cases
-	data := store.NewMap()
+	data := newStore()
 	// initialize the policy (with a recorder wrapping if logging is enabled)
-	var policy Policy = config.Policy(config.CacheSize, data)
+	policy := newPolicy(config.NumCounters, config.MaxCost)
 	if config.Log {
-		policy = NewRecorder(config.Policy(config.CacheSize, data), data)
+		policy = NewRecorder(policy, data)
 	}
 	return &Cache{
 		data:   data,
 		policy: policy,
-		buffer: ring.NewBuffer(ring.LOSSY, &ring.Config{
+		buffer: newRingBuffer(ringLossy, &ringConfig{
 			Consumer: policy,
-			Capacity: config.BufferSize,
+			Capacity: config.BufferItems,
 		}),
+		notify: config.OnEvict,
+		size:   config.NumCounters,
 	}
 }
 
-/*
-// TODO: finalize this and make it atomic
-func (c *Cache) GetOrSet(key string, get func() interface{}) interface{} {
-	value := c.Get(key)
-	if value == nil {
-		value = get()
-		c.Set(key, value)
-	}
-	return value
-}
-*/
-
 func (c *Cache) Get(key string) interface{} {
-	c.buffer.Push(ring.Element(key))
+	c.buffer.Push(ringItem(key))
 	return c.data.Get(key)
 }
 
-func (c *Cache) Set(key string, value interface{}) {
-	// attempt to add the key to the admission/eviction policy
-	if victim, added := c.policy.Add(key); added {
-		// delete the victim if there was an eviction
-		if victim != "" {
-			c.data.Del(key)
+func (c *Cache) Set(key string, val interface{}, cost uint64) ([]string, bool) {
```

Comment:
Would it make more sense to evaluate through an attached function?

In Caffeine, we offer a `Weigher` function that computes an `int32` given the (key, value) pair. The is a nice property for computations, e.g. loading on a cache miss atomically. An explicit value upfront isn't known, so passing through is less useful in a concurrent setting.

The total `maximumWeight` of the cache is an `int64`. This avoids integer overflow and those using byte-sizes we expect them to scale appropriately. I'd make sure you have tests to if the sum exceeds `MaxUint64`.

Because we know upfront whether the cache is weight- (cost) or size-based, we code generate optimized versions of the cache entry. If the cache is size-bounded (weigher => 1) then we drop the weight field for a slimmer object. All of the logic remains the same, it is simply ignored by the entry factory when allocating the instance and its accessor returns the constant value.

Reply:
> Would it make more sense to evaluate through an attached function?

I think that's a great idea (reminds me of [Hyperbolic Caching](https://dl.acm.org/citation.cfm?id=3154690.3154738)) with the lazily evaluated priority function using item metadata as parameters. I'll have to ask @manishrjain about the origins of the cost parameter, and whether or not it would be advantageous to replace it with a function pointer. Right now the goal is just playing nicely with Badger. 

> The is a nice property for computations, e.g. loading on a cache miss atomically. An explicit value upfront isn't known, so passing through is less useful in a concurrent setting.

This is a good point. We don't have an atomic load-on-miss method right now, and if we added one the passed function would have to return both the value and cost. Do you think there's a benefit to decoupling the cost retrieval function from the value retrieval function?

> Because we know upfront whether the cache is weight- (cost) or size-based, we code generate optimized versions of the cache entry. If the cache is size-bounded (weigher => 1) then we drop the weight field for a slimmer object. All of the logic remains the same, it is simply ignored by the entry factory when allocating the instance and its accessor returns the constant value.

I'd like to add this optimization in the future. Perhaps rather than generating the code I'll just use another Policy implementation with less overhead in the case of a size-bounded configuration. I can reuse the shared logic between Policies.

Reply:
> with the lazily evaluated priority function using item metadata as parameters.

Note that the function is evaluated on insertion or update, not on eviction, obviously because you need to track the totals. The entry's size may change and that must be done via a cache write for proper accounting, e.g. if it grew and forces an eviction. In a concurrent setting the race conditions has to be handled, such as when shifting entries between LRU lists, which is harder when the hash-table and policy are decoupled under different locks.

> the passed function would have to return both the value and cost. Do you think there's a benefit to decoupling the cost retrieval function from the value retrieval function?

Yes, I think it is simpler since the cost evaluation will be uniform (e.g. `len(data)`). In a simple load-on-miss case it might be simple if coupled, you could also have other complex cases to support such as cache refresh (`reload(key, oldValue)`), bulk load optimizations (`getAll(keys)` => `loadAll(keys)` for missing entries), and ad hoc computations (`compute(key, (key, oldValue) -> newValue)`). You would otherwise litter the evaluation everywhere, which is more error prone.

Another example is an async cache, where the value is a future, the entry's size is zero while in-flight (not eligible for eviction). When the future completes then a callback updates the cache, causing the entry's size to be calculated and an eviction to be triggered. The user-facing apis are much simpler / uniform / orthogonal with a function, imho.